app/src/main/java/ch/epfl/sdp/GoogleApi.java

```diff
@@ -0,0 +1,96 @@
+package ch.epfl.sdp;
+
+import android.Manifest;
+import android.app.Activity;
+import android.content.pm.PackageManager;
+import android.location.Criteria;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+
+import androidx.core.app.ActivityCompat;
+
+import com.google.android.gms.maps.CameraUpdateFactory;
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.LatLng;
+import com.google.android.gms.maps.model.Marker;
+import com.google.android.gms.maps.model.MarkerOptions;
+
+public class GoogleApi implements MapApi {
+    public static double listenTime = 1000; // milliseconds
+    public static double listenDistance = 5; // meters
+
+    private Location currentLocation = new Location("init");
+    private LocationManager locationManager;
+    private Criteria criteria;
+    private String bestProvider;
+    private LocationListener locationListener;
+    private Marker marker;
+    private GoogleMap mMap;
+
+    public GoogleApi(LocationManager locationManager) {
+        this.locationManager = locationManager;
+
+        // setup bestProvider
+        criteria = new Criteria();
+        criteria.setAccuracy(Criteria.ACCURACY_FINE);
+        bestProvider = locationManager.getBestProvider(criteria, true);
+
+        locationListener = new LocationListener() {
+            @Override
+            public void onLocationChanged(Location latestLocation) {
+                // Called when a new location is found by the network location provider.
+                currentLocation = latestLocation;
+            }
+
+            @Override
+            public void onStatusChanged(String provider, int status, Bundle extras) {}
+
+            @Override
+            public void onProviderEnabled(String provider) {}
+
+            @Override
+            public void onProviderDisabled(String provider) {}
+        };
+    }
+
+    @Override
+    public Location getCurrentLocation() {
+        return currentLocation;
+    }
+
+    @Override
+    public void updatePosition(Activity activity) {
+        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&

```

Comment:
Maybe parts of this method could be modularized somewhere else? It would enable you to test it, as you can probably not test this class because it uses Google Maps.